contracts/Controller.sol

```diff
@@ -495,13 +508,64 @@ contract Controller is ReentrancyGuard, Ownable {
      * @dev only account owner or operator can withdraw long option from vault
      * @param _args MintArgs structure
      */
-    // function _burnOtoken(Actions.BurnArgs memory _args) internal {}
+    function _burnOtoken(Actions.BurnArgs memory _args)
+        internal
+        isAuthorized(msg.sender, _args.owner)
+        returns (MarginAccount.Vault memory)
+    {
+        require(checkVaultId(_args.owner, _args.vaultId), "Controller: invalid vault id");
+        require(_args.from == msg.sender, "Controller: burner address and msg.sender address mismatch");
+
+        OtokenInterface otoken = OtokenInterface(_args.otoken);
+
+        require(now <= otoken.expiryTimestamp(), "Controller: can not burn expired otoken");
+
+        vaults[_args.owner][_args.vaultId]._removeShort(_args.otoken, _args.amount, _args.index);
+
+        otoken.burnOtoken(_args.from, _args.amount);
+
+        emit ShortOtokenBurned(_args.otoken, _args.owner, _args.from, _args.vaultId, _args.amount);
+
+        return vaults[_args.owner][_args.vaultId];
+    }
 
     /**
      * @notice exercise option
      * @param _args ExerciseArgs structure
      */
-    // function _exercise(Actions.ExerciseArgs memory _args) internal {}
+    function _exercise(Actions.ExerciseArgs memory _args) internal {
+        OtokenInterface otoken = OtokenInterface(_args.otoken);
+
+        require(now > otoken.expiryTimestamp(), "Controller: can not exercise un-expired otoken");
+
+        require(isPriceFinalized(_args.otoken), "Controller: otoken underlying asset price is not finalized yet");
+
+        uint256 payout = getPayout(_args.otoken, _args.amount);
+
+        otoken.burnOtoken(msg.sender, _args.amount);
+
+        address marginPoolModule = AddressBookInterface(addressBook).getMarginPool();
+        MarginPoolInterface marginPool = MarginPoolInterface(marginPoolModule);
+
+        marginPool.transferToUser(otoken.collateralAsset(), _args.exerciser, payout);
+
+        emit Exercise(_args.otoken, msg.sender, _args.exerciser, otoken.collateralAsset(), _args.amount, payout);
+    }
+
+    /**
+     * @notice get Otoken payout after expiry
+     * @param _otoken Otoken address
+     * @param _amount amount of Otoken
+     * @return payout = cashValue * amount
+     */
+    function getPayout(address _otoken, uint256 _amount) internal view returns (uint256) {
```

Comment:
```suggestion
    function _getPayout(address _otoken, uint256 _amount) internal view returns (uint256) {
```

Reply:
or consider making this public?

Reply:
why ? can be calculated by calling the getCashValue which is public and multiplying by the amount.

Reply:
just thought i could be handy but don't need to. 
if you want to keep it internal then add a `_`